File: 98. Validate Binary Search Tree/Solution.java
class Solution {
    public boolean isValidBST(TreeNode root) {
        return validate(root, null, null);
    }

    private boolean validate(TreeNode node, TreeNode min, TreeNode max) {
        if (node == null) {
            return true;
        }

        if ((max != null && node.val >= max.val) || (min != null && node.val <= min.val)) {
            return false;
        }

        return validate(node.left, min, node) && validate(node.right, node, max);
    }
}

File: 98. Validate Binary Search Tree/solution.js
function isNil(element) {
    return (element === null || element === undefined);
}

var isValidBST = function(root) {
    return validate(root);
}

function validate(node, min, max) {
    if (isNil(node)) {
        return true;
    }

    if (!isNil(max) && node.val >= max.val) {
        return false;
    }

    if (!isNil(min) && node.val <= min.val) {
        return false;
    }

    return validate(node.left, min, node) && validate(node.right, node, max);
}